File: leetcode-cn/24.两两交换链表中的节点.py
#
# @lc app=leetcode.cn id=24 lang=python3
#
# [24] 两两交换链表中的节点
#


# @lc code=start
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if head is None:
            return None

        ans = ListNode(-1)
        ans.next = head
        p = ans
        q = p.next
        r = q.next

        while p and q and r:
            p.next = r
            q.next = r.next
            r.next = q

            p = q
            if p:
                q = p.next
            else:
                q = None

            if q:
                r = q.next
            else:
                r = None

        return ans.next


# @lc code=end

File: leetcode-cn/C++/24.两两交换链表中的节点.cpp
/*
 * @lc app=leetcode.cn id=24 lang=cpp
 *
 * [24] 两两交换链表中的节点
 */

// @lc code=start
// struct ListNode {
//     int val;
//     ListNode* next;
//     ListNode(int x)
//         : val(x)
//         , next(NULL)
//     {
//     }
// };

class Solution
{
public:
    ListNode *swapPairs(ListNode *head)
    {
        if (head == nullptr)
        {
            return head;
        }

        ListNode *new_head = new ListNode();
        new_head->next = head;

        ListNode *pre = new_head;

        while (pre->next && pre->next->next)
        {
            ListNode *p = pre->next;
            ListNode *q = pre->next->next;
            pre->next = q;
            p->next = q->next;
            q->next = p;

            pre = p;
        }

        return new_head->next;
    }
};
// @lc code=end
